extra/DEVELOPER_DOCS/StringStore-Vocab.md

```diff
@@ -0,0 +1,220 @@
+# StringStore & Vocab
+
+## Overview
+
+spaCy represents mosts strings internally using a `uint64` in Cython which
+corresponds to a hash. The magic required to make this largely transparent is
+handled by the `StringStore`, and is integrated into the pipelines using the
+`Vocab`, which also connects it to some other information. 
+
+These are mostly internal details that average library users should never have
+to think about. On the other hand, when developing a component it's normal to
+interact with the Vocab for lexeme data or word vectors, and it's not unusual
+to add labels to the `StringStore`.
+
+## StringStore
+
+### Overview
+
+The `StringStore` is a `cdef class` that looks a bit like a two-way dictionary,
+though it is not a subclass of anything in particular.
+
+The main functionality of the `StringStore` is that `__getitem__` converts
+hashes into strings or strings into hashes. The details of this are very messy
+but generally not a problem in practice.
+
+Example:
+
+```
+from spacy.strings import `StringStore`
+
+ss = `StringStore`()
+hashval = ss["spacy"] # 10639093010105930009
+try:
+  # this won't work
+  ss[hashval]
+except:
+  pass
```

Comment:
```suggestion
except KeyError:
  print(f"key {hashval} unknown in the StringStore.")
```
I would also format this code block with 4 spaces indentation instead of 2 to mimic the normal `black` formatting we use throughout the code base.